core/vm/logger.go

```diff
@@ -335,3 +334,92 @@ func (t *mdLogger) CaptureEnd(output []byte, gasUsed uint64, tm time.Duration, e
 	fmt.Fprintf(t.out, "\nOutput: `0x%x`\nConsumed gas: `%d`\nError: `%v`\n",
 		output, gasUsed, err)
 }
+
+type accessList struct {
+	addresses map[common.Address]int
+	slots     []map[common.Hash]struct{}
+}
+
+func (al *accessList) AddAddress(address common.Address) {
+	// Set address if not previously present
+	if _, present := al.addresses[address]; !present {
+		al.addresses[address] = -1
+	}
+}
+
+func (al *accessList) AddSlot(address common.Address, slot common.Hash) {
+	idx, addrPresent := al.addresses[address]
+	if !addrPresent || idx == -1 {
+		// Address not present, or addr present but no slots there
+		al.addresses[address] = len(al.slots)
+		slotmap := map[common.Hash]struct{}{slot: {}}
+		al.slots = append(al.slots, slotmap)
+	}
+	// There is already an (address,slot) mapping
+	slotmap := al.slots[idx]
+	if _, ok := slotmap[slot]; !ok {
+		slotmap[slot] = struct{}{}
+	}
+	// No changes require
+}
+
+type AccessListTracer struct {
+	old  *types.AccessList
+	list accessList
+	err  error
+}
+
+func NewAccessListTracer(list *types.AccessList) *AccessListTracer {
+	return &AccessListTracer{
+		old: list,
+	}
+}
+
+func (a *AccessListTracer) CaptureStart(env *EVM, from common.Address, to common.Address, create bool, input []byte, gas uint64, value *big.Int)  {
+	env.StateDB.PrepareAccessList(from, &to, env.ActivePrecompiles(), *a.old)
+}
+
+func (a *AccessListTracer) CaptureState(env *EVM, pc uint64, op OpCode, gas, cost uint64, memory *Memory, stack *Stack, rData []byte, contract *Contract, depth int, err error)  {´
+	if (op == SLOAD || op == SSTORE) && stack.len() >= 1 {
+		slot := common.Hash(stack.data[stack.len()-1].Bytes32())
+		a.list.AddSlot(contract.Address(), slot)
+	}
+	if (op == EXTCODECOPY || op == EXTCODEHASH || op == EXTCODESIZE || op == BALANCE || op == SELFDESTRUCT) && stack.len() >= 1 {
+		stackElem := stack.data[stack.len()-1].Bytes32()
+		address := common.BytesToAddress(stackElem[:])
+		a.list.AddAddress(address)
+	}
+	if (op == DELEGATECALL || op == CALL || op == STATICCALL || op == CALLCODE) && stack.len() >= 5 {
+		stackElem := stack.data[stack.len()-5].Bytes32()
```

Comment:
The arguments are `gas, addess, ..`, so I'd expect the stackElem to be the second from the end, not the fifth?